third_party/nvfuser/csrc/scheduler/matmul_utils.cpp

```diff
@@ -0,0 +1,596 @@
+      const auto& in_B =
+          getShape(dynamic_cast<TensorView*>(fusion->inputs()[1]));
+      const auto& output =
+          getShape(dynamic_cast<TensorView*>(fusion->outputs()[0]));
+
+      constexpr size_t expected_dims = 2;
+      if (in_A.size() != expected_dims || //
+          in_B.size() != expected_dims || //
+          output.size() != expected_dims) {
+        return c10::nullopt;
+      }
+
+      switch (matmul_layout) {
+        case MatmulLayout::TT: {
+          // in_A := [K, M]
+          // in_B := [N, K]
+          // output := [N, M]
+          const bool check_k = in_A[0] == in_B[1];
+          const bool check_m = in_A[1] == output[1];
+          const bool check_n = in_B[0] == output[0];
+          if (!(check_k && check_m && check_n)) {
+            return c10::nullopt;
+          }
+          // [M, N, K]
+          return TensorShape{in_A[1], in_B[0], in_A[0]};
+        }
+        case MatmulLayout::NT: {
+          // in_A := [M, K]
+          // in_B := [N, K]
+          // output := [N, M]
+          const bool check_k = in_A[1] == in_B[1];
+          const bool check_m = in_A[0] == output[1];
+          const bool check_n = in_B[0] == output[0];
+          if (!(check_k && check_m && check_n)) {
+            return c10::nullopt;
+          }
+          // [M, N, K]
+          return TensorShape{in_A[0], in_B[0], in_A[1]};
+        }
+        case MatmulLayout::TN: {
+          // in_A := [K, M]
+          // in_B := [K, N]
+          // output := [N, M]
+          const bool check_k = in_A[0] == in_B[0];
+          const bool check_m = in_A[1] == output[1];
+          const bool check_n = in_B[1] == output[0];
+          if (!(check_k && check_m && check_n)) {
+            return c10::nullopt;
+          }
+          // [M, N, K]
+          return TensorShape{in_A[1], in_B[1], in_A[0]};
+        }
+        default:
+          return c10::nullopt;
+      }
+      break;
+    }
+    default:
+      break;
+  }
+  return c10::nullopt;
+}
+
+//! A helper for checking in depth the basic pattern (C=AxB) in provided fusion
+//! and MMA op defined in it.
+std::string checkMatmulPatternBasic(const MmaOp* mma_expr, Fusion* fusion) {
+  const auto fusion_inputs = fusion->inputs();
+  const auto fusion_outputs = fusion->outputs();
+  const auto mma_inputs = mma_expr->inputs();
+  const auto mma_outputs = mma_expr->outputs();
+
+  // Fusion input A: fusion_inputs[0]
+  // Fusion input B: fusion_inputs[1]
+  // Fusion output C: fusion_outputs[0]
+  // MMA input A: mma_inputs[0]
+  // MMA input B: mma_inputs[1]
+  // MMA output C: mma_outputs[0]
+
+  // Check dependencies:
+  // #1 - fusion [A,B] -> mma A
+  // #2 - fusion [B,A] -> mma B
+  // #3 - mma C -> fusion C
+  // #4 - fusion A, B, C -> global memory
+  // #5 - mma A, B -> local memory
+  // #6 - mma C -> global memory
+
+  // #1
+  auto path_to_mma_A =
+      DependencyCheck::getAllDependencyChains(fusion_inputs[0], mma_inputs[0]);
```

Comment:
maybe it'd be better to grab all paths from all fusion inputs to the mma_inputs[0]. Then you could check if the mma_inputs[0] or mma_inputs[1] is in that set. Similar for output checks. You could also make sure there's no intersection between the sets i.e. no intersection of the sets {dependency check from fusion inputs to the first mma input, and the fusion inputs to the second mma input}. Similar thing could be checked for epilogue, this might help you isolate the different portions of the fusion.

Reply:
@csarofeen I will check your idea with goal of creating a helper tool that can be reused in future for other matmul patterns. There is one thing that I will need to check - how dependencies between inputs look if fusion were to represent `C = alpha * A x B` set of expressions.

I expect that in scenario there will be intersection in paths to mma 1st input between fusion 1st input (alpha) and its 2nd input (A)

Reply:
Yeah, only recommendation I have is thoroughly read through iter_visit.cpp and scheduler/registry.cpp . Those have a lot of very specific graph checking logic and should give you a feel for some different ways to think/look through the graph.

For `C = alpha * A x B` I would think you'd just look through the history of A and check that the only branches are from scalar expressions otherwise you're on the path to the fusion input leading to A.

Reply:
I will revisit files you mentioned as I now understand that my approach is a little bit too strict / not abstract enough. I realized this when I though how to handle alpha. I focused too much on a single pattern even though this functionality could be expressed in the following ways:
```
C = (alpha * A) x B
C = A x (alpha * B)
C = alpha * (A x B)
```
so the current compile-time check needs redoing.

Thanks for pointing this out!